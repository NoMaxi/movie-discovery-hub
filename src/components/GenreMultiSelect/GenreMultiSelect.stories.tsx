import type { Meta, StoryObj } from "@storybook/react";
import { GenreMultiSelect } from "./GenreMultiSelect";

const meta = {
    title: "Components/GenreMultiSelect",
    component: GenreMultiSelect,
    argTypes: {
        id: {
            control: "text",
            description: "Unique identifier for the select component",
        },
        name: {
            control: "text",
            description: "Name attribute for the hidden input field",
        },
        preselectedGenres: {
            control: "object",
            description: "Array of preselected genres",
        },
        ariaDescribedby: {
            control: "text",
            description: "ID of the element that describes this select for accessibility",
        },
    },
    args: {
        id: "genre-select",
        name: "genres",
    },
} satisfies Meta<typeof GenreMultiSelect>;

export default meta;

type Story = StoryObj<typeof meta>;

export const NoSelectedGenres: Story = {
    args: {},
};

export const PreselectedGenres: Story = {
    args: {
        preselectedGenres: ["Comedy", "Horror"],
    },
};

export const AllGenresPreselected: Story = {
    args: {
        preselectedGenres: ["Comedy", "Crime", "Documentary", "Horror"],
    },
};

export const WithAriaDescription: Story = {
    args: {
        preselectedGenres: ["Comedy"],
        ariaDescribedby: "genre-description",
    },
    decorators: [
        (Story) => (
            <div>
                <Story />
                <div id="genre-description" className="mb-2 text-sm text-gray-500">
                    Select one or more movie genres
                </div>
            </div>
        ),
    ],
};
