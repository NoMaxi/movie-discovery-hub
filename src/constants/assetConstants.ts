import noPosterImage from "@/assets/no-poster-image.png";

export const DEFAULT_IMAGE_URL = noPosterImage;
