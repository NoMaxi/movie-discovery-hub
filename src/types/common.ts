export type Genre = "All" | "Documentary" | "Comedy" | "Horror" | "Crime";

export interface Movie {
    id: string;
    imageUrl: string;
    title: string;
    releaseYear: number;
    genres: string[];
}
