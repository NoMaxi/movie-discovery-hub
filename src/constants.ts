import { Genre, SelectableGenre, SortOption } from "@/types/common";

export const PRIMARY_COLOR = "#F65261";

export const GENRES: Genre[] = ["All", "Documentary", "Comedy", "Horror", "Crime"];
export const SELECTABLE_GENRES: SelectableGenre[] = ["Documentary", "Comedy", "Horror", "Crime"];

export const DEFAULT_SORT_CRITERION: SortOption = "Release Date";
export const DEFAULT_ACTIVE_GENRE: Genre = "All";
export const DEFAULT_MOVIES_PER_PAGE = 18;
