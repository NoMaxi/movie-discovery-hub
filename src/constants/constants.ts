import { Genre, SelectableGenre, SortOption } from "@/types/common";

export const GENRES: Genre[] = ["All", "Documentary", "Comedy", "Horror", "Crime"];
export const SELECTABLE_GENRES: SelectableGenre[] = ["Documentary", "Comedy", "Horror", "Crime"];

export const PRIMARY_COLOR = "#F65261";

export const DEFAULT_SORT_CRITERION: SortOption = "Release Date";
export const DEFAULT_ACTIVE_GENRE: Genre = "All";
export const DEFAULT_TITLE = "No Title";
export const DEFAULT_DESCRIPTION = "No description available.";
export const DEFAULT_MOVIES_PER_PAGE = 18;
export const MOVIE_LIST_STALE_TIME = 5 * 60 * 1000; // 5 minutes
