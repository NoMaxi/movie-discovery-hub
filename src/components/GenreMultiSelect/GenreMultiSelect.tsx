import { forwardRef, useCallback, useEffect, useRef, useState } from "react";
import { Genre } from "@/types/common";
import { SELECTABLE_GENRES } from "@/constants/constants";
import { useClickOutside } from "@/hooks/useClickOutside/useClickOutside";
import SelectArrow from "@/components/common/SelectArrow/SelectArrow";

export interface GenreMultiSelectProps {
    id: string;
    name: string;
    preselectedGenres?: string[];
    onChange?: (value: string[]) => void;
    ariaDescribedby?: string;
    error?: boolean;
}

const defaultEmptyGenres: string[] = [];

export const GenreMultiSelect = forwardRef<HTMLDivElement, GenreMultiSelectProps>(
    ({ preselectedGenres = defaultEmptyGenres, onChange, id, ariaDescribedby, error }, ref) => {
        const [isOpen, setIsOpen] = useState(false);
        const [selectedGenres, setSelectedGenres] = useState<Set<Genre | string>>(
            () => new Set(preselectedGenres),
        );
        const controlRef = useRef<HTMLDivElement>(null);

        useEffect(() => {
            setSelectedGenres(new Set(preselectedGenres));
        }, [preselectedGenres]);

        useClickOutside(controlRef, () => setIsOpen(false), isOpen);

        const toggleDropdown = () => setIsOpen(!isOpen);

        const handleCheckboxChange = useCallback(
            (genre: Genre, isChecked: boolean) => {
                const newSelected = new Set(selectedGenres);
                if (isChecked) {
                    newSelected.add(genre);
                } else {
                    newSelected.delete(genre);
                }

                const genresArray = Array.from(newSelected);
                setSelectedGenres(newSelected);
                onChange?.(genresArray);
            },
            [selectedGenres, onChange],
        );

        const areGenresSelected = selectedGenres.size > 0;
        const displayValue = areGenresSelected ? Array.from(selectedGenres).join(", ") : "Select Genre";

        const combinedRef = (node: HTMLDivElement | null) => {
            controlRef.current = node;
            if (typeof ref === "function") {
                ref(node);
            } else if (ref) {
                ref.current = node;
            }
        };

        return (
            <div className="genre-select relative" ref={combinedRef}>
                <button
                    id={id}
                    type="button"
                    onClick={toggleDropdown}
                    aria-haspopup="listbox"
                    aria-expanded={isOpen}
                    aria-describedby={ariaDescribedby}
                    aria-invalid={error ? "true" : "false"}
                    className={`
                        input-field relative flex items-center justify-between w-full
                        text-left text-[var(--color-primary)] font-normal opacity-80
                        cursor-pointer ${error ? "input-error" : ""}
                    `}
                >
                    <span
                        className={`truncate ${
                            areGenresSelected ? "text-[var(--color-text)]" : "text-[var(--color-text)] opacity-30"
                        }`}
                    >
                        {displayValue}
                    </span>
                    <span className={`transform transition-transform ${isOpen ? "rotate-180" : "rotate-0"}`}>
                        <SelectArrow className="w-4 h-4" />
                    </span>
                </button>

                {isOpen && (
                    <div
                        role="listbox"
                        className="
                        absolute z-30 mt-1 w-full bg-[var(--color-background)]
                        border border-[var(--color-gray-light)] rounded-md shadow-lg
                        max-h-60 overflow-y-auto
                    "
                    >
                        {SELECTABLE_GENRES.map((genre) => (
                            <label
                                key={genre}
                                className="
                                    flex items-center px-4 py-3 space-x-3
                                    hover:bg-[var(--color-gray)]
                                    cursor-pointer transition-colors
                                "
                            >
                                <input
                                    type="checkbox"
                                    className="form-checkbox"
                                    checked={selectedGenres.has(genre)}
                                    onChange={(e) => handleCheckboxChange(genre, e.target.checked)}
                                    onClick={(e) => e.stopPropagation()}
                                />
                                <span className="text-[var(--color-text)] text-[16px]">{genre}</span>
                            </label>
                        ))}
                    </div>
                )}
            </div>
        );
    },
);
