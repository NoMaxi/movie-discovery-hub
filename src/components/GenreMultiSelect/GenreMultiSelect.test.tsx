import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent, { UserEvent } from "@testing-library/user-event";
import "@testing-library/jest-dom";
import { Genre } from "@/types/common";
import { SELECTABLE_GENRES } from "@/constants/constants";
import { GenreMultiSelect } from "./GenreMultiSelect";

jest.mock("@/hooks/useClickOutside/useClickOutside", () => ({
    useClickOutside: jest.fn(),
}));

jest.mock("@/components/common/SelectArrow/SelectArrow", () => ({
    __esModule: true,
    default: (props: React.SVGProps<SVGSVGElement>) => <svg data-testid="select-arrow" {...props} />,
}));

describe("GenreMultiSelect", () => {
    const defaultProps: React.ComponentProps<typeof GenreMultiSelect> = {
        id: "genre-select",
        name: "genres",
    };
    let user: UserEvent;

    beforeEach(() => {
        user = userEvent.setup();
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    const renderComponent = (props: Partial<React.ComponentProps<typeof GenreMultiSelect>> = {}) => {
        const ref = React.createRef<HTMLDivElement>();
        const mergedProps = { ...defaultProps, ...props };
        const view = render(<GenreMultiSelect {...mergedProps} ref={ref} />);
        return { ...view, ref };
    };

    test("Should properly handle function refs", () => {
        const mockRef = jest.fn();
        render(<GenreMultiSelect {...defaultProps} ref={mockRef} />);

        expect(mockRef).toHaveBeenCalledWith(expect.any(HTMLDivElement));
    });

    test("Should render correctly with default props and no preselected genres", () => {
        const { asFragment } = renderComponent();

        expect(screen.getByRole("button")).toBeInTheDocument();
        expect(screen.getByRole("button")).toHaveTextContent("Select Genre");
        expect(screen.getByTestId("select-arrow")).toBeInTheDocument();
        expect(screen.queryByRole("listbox")).not.toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    test("Should render correctly with preselected genres", () => {
        const preselectedGenres: Genre[] = ["Comedy", "Horror"];
        const { asFragment } = renderComponent({ preselectedGenres });

        expect(screen.getByRole("button")).toHaveTextContent("Comedy, Horror");
        expect(asFragment()).toMatchSnapshot();
    });

    test("Should toggle dropdown visibility when button is clicked", async () => {
        renderComponent();
        const button = screen.getByRole("button");

        expect(screen.queryByRole("listbox")).not.toBeInTheDocument();

        await user.click(button);

        expect(screen.getByRole("listbox")).toBeInTheDocument();

        SELECTABLE_GENRES.forEach((genre) => {
            expect(screen.getByText(genre)).toBeInTheDocument();
            expect(screen.getByLabelText(genre, { selector: "input[type='checkbox']" })).not.toBeChecked();
        });

        await user.click(button);

        expect(screen.queryByRole("listbox")).not.toBeInTheDocument();
    });

    test("Should update selected genres when checkboxes are clicked", async () => {
        renderComponent();
        const button = screen.getByRole("button");

        await user.click(button);

        const comedyCheckbox = screen.getByLabelText("Comedy", { selector: "input[type='checkbox']" });
        const horrorCheckbox = screen.getByLabelText("Horror", { selector: "input[type='checkbox']" });

        await user.click(comedyCheckbox);

        expect(comedyCheckbox).toBeChecked();
        expect(screen.getByRole("button")).toHaveTextContent("Comedy");

        await user.click(horrorCheckbox);

        expect(horrorCheckbox).toBeChecked();
        expect(screen.getByRole("button")).toHaveTextContent("Comedy, Horror");

        await user.click(comedyCheckbox);

        expect(comedyCheckbox).not.toBeChecked();
        expect(horrorCheckbox).toBeChecked();
        expect(screen.getByRole("button")).toHaveTextContent("Horror");
    });

    test("Should display correct checkboxes state with preselected genres", async () => {
        const preselectedGenres: Genre[] = ["Comedy", "Crime"];
        renderComponent({ preselectedGenres });
        const button = screen.getByRole("button");

        await user.click(button);

        const comedyCheckbox = screen.getByLabelText("Comedy", { selector: "input[type='checkbox']" });
        const crimeCheckbox = screen.getByLabelText("Crime", { selector: "input[type='checkbox']" });
        const documentaryCheckbox = screen.getByLabelText("Documentary", { selector: "input[type='checkbox']" });
        const horrorCheckbox = screen.getByLabelText("Horror", { selector: "input[type='checkbox']" });

        expect(comedyCheckbox).toBeChecked();
        expect(crimeCheckbox).toBeChecked();
        expect(documentaryCheckbox).not.toBeChecked();
        expect(horrorCheckbox).not.toBeChecked();
    });

    test("Should handle selections and trigger onChange callback", async () => {
        const onChange = jest.fn();
        renderComponent({ onChange });
        const button = screen.getByRole("button");

        await user.click(button);

        const comedyCheckbox = screen.getByLabelText("Comedy", { selector: "input[type='checkbox']" });
        const documentaryCheckbox = screen.getByLabelText("Documentary", { selector: "input[type='checkbox']" });

        await user.click(comedyCheckbox);
        expect(onChange).toHaveBeenCalledWith(["Comedy"]);

        await user.click(documentaryCheckbox);
        expect(onChange).toHaveBeenCalledWith(["Comedy", "Documentary"]);

        await user.click(documentaryCheckbox);
        expect(onChange).toHaveBeenCalledWith(["Comedy"]);
    });

    test("Should apply accessibility attributes correctly", () => {
        const ariaDescribedby = "test-description";
        renderComponent({ ariaDescribedby });
        const button = screen.getByRole("button");

        expect(button).toHaveAttribute("aria-haspopup", "listbox");
        expect(button).toHaveAttribute("aria-expanded", "false");
        expect(button).toHaveAttribute("aria-describedby", ariaDescribedby);
        expect(button).toHaveAttribute("id", defaultProps.id);
    });

    test("Should update aria-expanded attribute when dropdown is toggled", async () => {
        renderComponent();
        const button = screen.getByRole("button");

        expect(button).toHaveAttribute("aria-expanded", "false");

        await user.click(button);

        expect(button).toHaveAttribute("aria-expanded", "true");

        await user.click(button);

        expect(button).toHaveAttribute("aria-expanded", "false");
    });

    test("Should update selected genres when preselectedGenres prop changes", async () => {
        const { rerender } = renderComponent({ preselectedGenres: ["Comedy"] });

        expect(screen.getByRole("button")).toHaveTextContent("Comedy");

        rerender(<GenreMultiSelect {...defaultProps} preselectedGenres={["Horror"]} />);

        expect(screen.getByRole("button")).toHaveTextContent("Horror");

        rerender(<GenreMultiSelect {...defaultProps} preselectedGenres={[]} />);

        expect(screen.getByRole("button")).toHaveTextContent("Select Genre");
    });

    test("Should apply error styles and aria-invalid when error prop is true", () => {
        renderComponent({ error: true });
        const button = screen.getByRole("button");

        expect(button).toHaveClass("input-error");
        expect(button).toHaveAttribute("aria-invalid", "true");
    });

    test("Should prevent event propagation when clicking on checkboxes", async () => {
        renderComponent();
        const button = screen.getByRole("button");

        await user.click(button);

        const comedyCheckbox = screen.getByLabelText("Comedy", { selector: "input[type='checkbox']" });

        const mockStopPropagation = jest.fn();
        const originalAddEventListener = document.addEventListener;

        document.addEventListener = jest.fn().mockImplementation((event, handler) => {
            if (event === "click") {
                document.addEventListener = originalAddEventListener;
            } else {
                originalAddEventListener(event, handler);
            }
        });

        const clickEvent = new MouseEvent("click", { bubbles: true });
        Object.defineProperty(clickEvent, "stopPropagation", { value: mockStopPropagation });

        await user.click(comedyCheckbox);

        expect(screen.getByRole("listbox")).toBeInTheDocument();

        document.addEventListener = originalAddEventListener;
    });
});
