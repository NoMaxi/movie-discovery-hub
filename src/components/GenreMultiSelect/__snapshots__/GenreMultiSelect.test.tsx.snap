// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenreMultiSelect Should render correctly with default props and no preselected genres 1`] = `
<DocumentFragment>
  <div
    class="genre-select relative"
  >
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-invalid="false"
      class="
                        input-field relative flex items-center justify-between w-full
                        text-left text-[var(--color-primary)] font-normal opacity-80
                        cursor-pointer 
                    "
      id="genre-select"
      type="button"
    >
      <span
        class="truncate text-[var(--color-text)] opacity-30"
      >
        Select Genre
      </span>
      <span
        class="transform transition-transform rotate-0"
      >
        <svg
          class="w-4 h-4"
          data-testid="select-arrow"
        />
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`GenreMultiSelect Should render correctly with preselected genres 1`] = `
<DocumentFragment>
  <div
    class="genre-select relative"
  >
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-invalid="false"
      class="
                        input-field relative flex items-center justify-between w-full
                        text-left text-[var(--color-primary)] font-normal opacity-80
                        cursor-pointer 
                    "
      id="genre-select"
      type="button"
    >
      <span
        class="truncate text-[var(--color-text)]"
      >
        Comedy, Horror
      </span>
      <span
        class="transform transition-transform rotate-0"
      >
        <svg
          class="w-4 h-4"
          data-testid="select-arrow"
        />
      </span>
    </button>
  </div>
</DocumentFragment>
`;
